cdc/puller/multiplexing_puller.go

```diff
@@ -0,0 +1,428 @@
+		PullerEventCounter.DeleteLabelValues(p.changefeed.Namespace, p.changefeed.ID, "resolved")
+		pullerQueueDuration.DeleteLabelValues(p.changefeed.Namespace, p.changefeed.ID, "kv")
+		pullerQueueDuration.DeleteLabelValues(p.changefeed.Namespace, p.changefeed.ID, "resolved")
+	}()
+
+	g, ctx := errgroup.WithContext(ctx)
+	if includeClient {
+		g.Go(func() error { return p.client.Run(ctx) })
+	}
+
+	g.Go(func() error { return p.checkResolveLock(ctx) })
+
+	for i := 0; i < p.frontiers; i++ {
+		g.Go(func() error { return p.advanceSpans(ctx) })
+	}
+	for i := range p.inputChs {
+		inputCh := p.inputChs[i]
+		g.Go(func() error { return p.handleInputCh(ctx, inputCh) })
+	}
+
+	log.Info("MultiplexingPuller starts",
+		zap.String("namespace", p.changefeed.Namespace),
+		zap.String("changefeed", p.changefeed.ID),
+		zap.Int("frontierConcurrent", p.frontiers))
+	return g.Wait()
+}
+
+func (p *MultiplexingPuller) handleInputCh(ctx context.Context, inputCh <-chan kv.MultiplexingEvent) error {
+	for {
+		var e kv.MultiplexingEvent
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case e = <-inputCh:
+		}
+
+		progress := p.getProgress(e.SubscriptionID)
+		if progress == nil {
+			continue
+		}
+
+		if e.Val != nil {
+			p.queueKvDuration.Observe(float64(time.Since(e.Start).Milliseconds()))
+			p.pullerEventCounterKv.Inc()
+			if err := progress.consume.f(ctx, e.Val, progress.spans); err != nil {
+				return errors.Trace(err)
+			}
+		} else if e.Resolved != nil {
+			p.pullerEventCounterResolved.Add(float64(len(e.Resolved.Spans)))
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case progress.resolvedSpans <- e:
+				p.schedule(ctx, progress)
+			}
+		}
+	}
+}
+
+func (p *MultiplexingPuller) getProgress(subID kv.SubscriptionID) *tableProgress {
+	p.subscriptions.RLock()
+	defer p.subscriptions.RUnlock()
+	return p.subscriptions.m[subID]
+}
+
+func (p *MultiplexingPuller) getAllProgresses() map[*tableProgress]struct{} {
+	p.subscriptions.RLock()
+	defer p.subscriptions.RUnlock()
+	hashset := make(map[*tableProgress]struct{}, len(p.subscriptions.m))
+	for _, value := range p.subscriptions.m {
+		hashset[value] = struct{}{}
+	}
+	return hashset
+}
+
+func (p *MultiplexingPuller) getProgressBySpan(span tablepb.Span) *tableProgress {
+	p.subscriptions.RLock()
+	defer p.subscriptions.RUnlock()
+	return p.subscriptions.n.GetV(span)
+}
+
+func (p *MultiplexingPuller) schedule(ctx context.Context, progress *tableProgress) {
+	if progress.scheduled.CompareAndSwap(false, true) {
+		select {
+		case <-ctx.Done():
+		case p.advanceCh <- progress:
+		}
+	}
+}
+
+func (p *MultiplexingPuller) advanceSpans(ctx context.Context) error {
+	handleProgress := func(ctx context.Context, progress *tableProgress) error {
+		defer func() {
+			progress.scheduled.Store(false)
+			if len(progress.resolvedSpans) > 0 {
+				p.schedule(ctx, progress)
+			}
+		}()
+
+		var event kv.MultiplexingEvent
+		var spans *model.ResolvedSpans
+		for i := 0; i < 128; i++ {
```

Comment:
Why handle resolvedTs events asynchronously? And what happens if consumption is slower here?

Reply:
By default we uses more threads to handle resolved timestamps. It's helpful for large tables because `Frontier` utilizes too much CPU.